Update README for Gateway tests (#14907)
kyma-project/kyma: tests/components/application-connector/README.md

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
**Table of Contents**

- [Component tests for Application Connector](#component-tests-for-application-connector)
- [Design and architecture](#design-and-architecture)
- [Building](#building)
- [Running](#running)
    - [Deploy Kyma cluster locally](#deploy-kyma-cluster-locally)
    - [Run the tests](#run-the-tests)
- [Debugging](#debugging)
    - [Running locally](#running-locally)
    - [Running without cleanup](#running-without-cleanup)

<!-- markdown-toc end -->

## Design and architecture

The tests consist of:
- [Application CRs](./resources/charts/gateway-test/templates/applications/) describing test cases
- [Test runners](./test/application-gateway/) with various check for subsets of cases, grouped by Application CRs
- [Mock application](./tools/external-api-mock-app/) which simulates remote endpoints

Additionally, following resources are created on the cluster:
- [Service Account](./resources/charts/gateway-test/templates/service-account.yml:2), used by tests to read Application CRs

The tests are executed as a Kubernetes Job on a Kyma cluster, 

> where the tested Application Gateway is installed. 

Author: Is there some middle ground between not splitting and not making 111 char lines?

- `{DOCKER_PUSH_REPOSITORY}/mock-app:{DOCKER_TAG}`

## Running

Tests can be run on any Kyma cluster with Application Gateway.

### Deploy Kyma cluster locally

1. Provision a local Kubernetes cluster with k3d:
   ```sh
   kyma provision k3d
   ```

1. Install the minimal set of components required to run Application Gateway for either Kyma OS or SKR:

    <div tabs name="Kyma flavor" group="minimal-kyma-installation">
    <details open>
    <summary label="OS">
    Kyma OS
    </summary>

    ```sh
    kyma deploy --components-file ./resources/installation-config/mini-kyma-os.yaml
    ```

    </details>
    <details>
    <summary label="SKR">
    SKR
    </summary>

    ```bash
    kyma deploy --components-file ./resources/installation-config/mini-kyma-skr.yaml 
    ```

    </details>
    </div>

    >**TIP:** More on Kyma installation can be found in [official Kyma documentation](https://kyma-project.io/docs/kyma/latest/02-get-started/01-quick-install/#install-kyma)

### Run the tests

``` sh
make test-gateway
```

By default tests clean up after themselves, removing all created resources
and `test` Namespace.

> **CAUTION:** This might override or remove existing resources, 

> > if names of already existing resources collide with names used by tests

Author: https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/02-style-and-terminology.md#tenses

> > if names of already existing resources collide with names used by tests

Author: > Question: is there a quick and easy way to check what are the names used by the tests?

Not really, at this moment, but those are constant, so we could just list them. Writing a script to list names also shouldn't take more than few minutes, but if we want to fix it, the best idea would be to add parametrized prefix to names

> > if names of already existing resources collide with names used by tests

Reviewer: I believe it would be helpful for the user to know what these resources are called in order to avoid naming collision, but I also don't think it has to be done on this PR. You might log it as a follow-up, for example. 

As for the tense, touche :) I edited the suggestion.

## Debugging

### Running locally

> The Test Job must run on a cluster, because of the way it accesses Application CRs.

Author: I do not explain how to run it locally and it'll just crash instantly, so it should be hard to break anything by doing it incorrectly, I don't know if it's caution-worthy. It just won't start

> The Test Job must run on a cluster, because of the way it accesses Application CRs.

Reviewer: I was thinking more along the lines of this being important enough to draw special attention to it -- it might not do any harm, but the user might not understand why it crashes when they run it locally and think that there's a problem with the tests, and not with the way they run them.
But it's just a thought. I'm okay with both ways.

Application Gateway and Mock Application can both be run locally.

To run Mock Application locally:
1. Change all `targetUrl` in [Application CRs](./resources/charts/gateway-test/templates/applications/)
   to reflect new app URL. For example `http://localhost:8081/v1/api/unsecure/ok`
1. Change all `centralGatewayUrl` to reflect new Application Gateway URL. 
   For example `http://localhost:8080/positive-authorisation/unsecure-always-ok`
1. Deploy all resources (you can omit test Job and Central Gateway, but it's easier to just let them fail)
   on the cluster

> 1. Build the Mock Application:

Reviewer: ```suggestion
4. Build the mock application:
```
- Be consistent with capitalization. 
- Use actual numbering.

   
   <div tabs name="Mock App Build Flavor" group="mock-app-flavor">
   <details open>
   <summary label="dockerized">
   Docker
   </summary>

   ```shell
   export DOCKER_TAG="local"

>    export DOCKER_PUSH_REPOSITORY="{Docker username}"

Reviewer: ```suggestion
   export DOCKER_PUSH_REPOSITORY="{DOCKER_USERNAME}"
```
We normally always write the placeholders in all caps and using the underscore `_` instead of space. See the [guidelines](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#use-bold-font-for-these-items) for reference.

   make image-mock-app
   ```

   </details>
   <details>
   <summary label="local">
   Local
   </summary>

>    Change hardcoded application port in [config.go](./tools/external-api-mock-app/config.go), then

Reviewer: ```suggestion
   Change the hardcoded application port in [`config.go`](./tools/external-api-mock-app/config.go), and run:
```
- Use articles.
- Format file names as `code`. See the [guidelines](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/03-formatting.md#use-code-font-for-these-items) for reference.
- Don't break a sentence half-way.

   ```shell
   go build ./tools/external-api-mock-app/
   ```
   </details>
   </div>

> 1. Run the Mock Application:

Reviewer: ```suggestion
5. Run the mock application:
```
- Be consistent with capitalization.
- Use actual numbering.

   
   <div tabs name="Mock App Run Flavor" group="mock-app-flavor">
   <details open>
   <summary label="dockerized">
   Docker
   </summary>

   ```shell
   docker run -p 8180:8080 -p 8190:8090 -v "$PWD/k8s/gateway-test/certs:/etc/secret-volume:ro" "$DOCKER_PUSH_REPOSITORY/mock-app:$DOCKER_TAG"
   ```

   </details>
   <details>
   <summary label="local">
   Local
   </summary>

   ```shell
   ./external-api-mock-app
   ```

>    > **CAUTION:** Certificates won't work, unless you copy them from `./k8s/gateway-test/certs` to `/etc/secret-volume`

Reviewer: ```suggestion
   > **CAUTION:** For the certificates to work, you must copy them from `./k8s/gateway-test/certs` to `/etc/secret-volume`.
```

Could we add a step that does that? Is this needed? 
Let's try to align on that on Monday if possible.

   </details>
   </div>

> 1. Run [Central Application Gateway](https://github.com/kyma-project/kyma/tree/main/components/central-application-gateway),
>    with `-kubeConfig {path_to_kubeconfig.yaml}` parameter.

Reviewer: ```suggestion
6. Run [Application Gateway](https://github.com/kyma-project/kyma/tree/main/components/central-application-gateway) with the `-kubeConfig {PATH_TO_YOUR_KUBECONFIG_FILE}` parameter.
```
- Use actual numbering. 
- You don't call this **Central** Application Gateway anywhere else in the document. Please unify the naming. 
- Don't split a sentence into two lines. 
- The placeholders are written in all caps. 
- What's more, the kubeconfig doesn't necessarily have to be called `kubeconfig.yaml`, right?

> You can now send requests to the Application Gateway and debug its behaviour locally.

Reviewer: ```suggestion
You can now send requests to Application Gateway, and debug its behavior locally.
```
Nice that you included the result :)

- We don't use the definite article with the names of our components. 
- `behaviour` is the BrE spelling, we use AmE (`behavior`) in our documentation. See the [guidelines](https://github.com/kyma-project/community/blob/main/docs/guidelines/content-guidelines/02-style-and-terminology.md#terminology) for reference.

### Running without cleanup

> To run the tests without removing all the resources:

Reviewer: ```suggestion
To run the tests without removing all the created resources afterwards, run:
```
- Don't break a sentence half-way (add a verb to finish it, then add the command).
- I added more precision to the sentence.
